Make metadata collection configurable

hush-hush opened the pull request:
### What does this PR do?

Make metadata collection configurable.

This also change the default interval for the `host` metadata provider (from once every 4 hours to once every 30min).

olivielpeau commented:
would it make sense to make this interval configurable under the existing `metadata_providers` key in `datadog.yaml`?

ogaca-dd commented on pkg/config/config.go:
```diff
@@ -119,6 +119,12 @@ func initConfig(config Config) {
+	// run the host metadata collector every 14400 seconds (4 hours)
+	config.BindEnvAndSetDefault("host_metadata_interval", 14400)
+	// run the agent checks metadata collector every 600 seconds (10 minutes)
+	config.BindEnvAndSetDefault("agent_check_metadata_internval", 600)
```
typo: inter--n--val

ogaca-dd reviewed (commented):
Is there any documentation, default config to update?

hush-hush commented:
> would it make sense to make this interval configurable under the existing `metadata_providers` key in `datadog.yaml`?

Those are the default metadata provider run by the agent no matter the configuration where `metadata_providers` is a list of optional provider. I think it make more sens to have those value outside `metadata_providers`.

olivielpeau commented:
> Those are the default metadata provider run by the agent no matter the configuration where metadata_providers is a list of optional provider. I think it make more sens to have those value outside metadata_providers.

You're right that these are default metadata providers, but we already have logic to handle this specific case for the `resources` provider [here](https://github.com/DataDog/datadog-agent/pull/3939/files#diff-1da27710e4127e711e81e8e4d4b390a5L273), we could re-use the same logic for the other default metadata providers.

Adding @ofek as reviewer as I think there was interest from the Integrations team to add this feature as well.

ofek commented:
FYI https://github.com/DataDog/datadog-agent/pull/3959

hush-hush committed:
Make metadata collection configurable

Default metadata collection can be configured within predefined range.
'host' default collection interval has been lowered to 30min by default.

olivielpeau commented on cmd/agent/app/start.go:
```diff
@@ -233,19 +242,29 @@ func StartAgent() error {
+	defaultCollectors := map[string]collector{
+		"host":         collector{os: "*", interval: hostMetadataCollectorInterval},
+		"agent_checks": collector{os: "*", interval: agentChecksMetadataCollectorInterval},
+		"resources":    collector{os: "linux", interval: resourcesMetadataCollectorInterval},
```
since `interval` is of type `time.Duration`, the conversion from `int` (in seconds) to `time.Duration` should be done here instead of when you call `common.MetadataScheduler.AddCollector`

(now that I write this, I realize the same mistake is present on the `MetadataProviders` struct definition. We should use `int` in that struct to match the config, so maybe for consistency the type of `collector.interval` defined in this function should be `int` as well)

olivielpeau replied:
(this comment also applies to the dogstatsd file)

olivielpeau commented on cmd/agent/app/start.go:
```diff
@@ -261,19 +280,16 @@ func setupMetadataCollection(s *serializer.Serializer, hostname string) error {
+		err = common.MetadataScheduler.AddCollector(name, c.interval*time.Second)
 		if err != nil {
-			log.Warn("Could not add resources metadata provider: ", err)
+			log.Error("Could not add resources metadata provider for %s: %v", name, err)
```
```suggestion
			log.Error("Could not add metadata provider for %s: %v", name, err)
```

olivielpeau commented on cmd/agent/app/start.go:
```diff
@@ -261,19 +280,16 @@ func setupMetadataCollection(s *serializer.Serializer, hostname string) error {
 		if err != nil {
-			log.Warn("Could not add resources metadata provider: ", err)
+			log.Error("Could not add resources metadata provider for %s: %v", name, err)
+			return err
```
note: with your changes, if there's an error when adding the `resources` metadata collector, we now return an error instead of returning `nil`. This ultimately means that the Agent would shut down if the `resources` metadata can't be initialized.

Could you check the implementation and build tags of the `resources` collector to confirm this won't make the Agent shut down on startup in some scenarios/builds where we'd actually want the Agent to ignore the error and continue running?

hush-hush replied:
I replicated the behavior with the new implementation.

olivielpeau commented on pkg/metadata/helper.go:
```diff
@@ -0,0 +1,114 @@
+
+var (
+	// default collectors by os
+	defaultCollectors = map[string]collector{
```
these are the default collectors of the Agent specifically (they don't really apply to Dogstatsd). How about defining them in `cmd/agent/`, and passing them to `SetupMetadataCollection`? I _think_ you could then make dogstatsd use `SetupMetadataCollection` as well, or am I missing something?

hush-hush replied:
The agent and dogstatsd share the same logic (same default interval, min, max ,...). I would like to avoid duplicating the data when we do want to keep the same behavior.

Plus I think this logic belongs in the `metadata` package and not outside. It's up to the `metadata` package to make we send the payloads in an healthy way.

And it's up to each binary to know which provider they want to start by default.

WDYT ?

olivielpeau replied:
I agree with both of your points, but to me it looks like the code in `cmd/dogstatsd` is doing too much compared to the code in `cmd/agent`. To me:

* the only thing the code in `cmd/dogstatsd` _should_ be doing is define that it only wants the `host` provider by default
* the only thing the code in `cmd/agent` should be doing is define that it wants "all" the default providers; or alternatively that it wants `host`, `agent_checks` and `resources`

olivielpeau commented on pkg/metadata/helper.go:
```diff
@@ -0,0 +1,131 @@
+		}
+		err := sch.AddCollector(name, cInfo.interval)
+		if err != nil && cInfo.ignoreError == false {
+			log.Errorf("Could not add resources metadata provider for %s: %v", name, err)
```
```suggestion
			log.Warnf("Could not add metadata provider for %s: %v", name, err)
```

olivielpeau commented on pkg/metadata/helper.go:
```diff
@@ -0,0 +1,131 @@
+}
+
+// SetupMetadataCollection initializes the metadata scheduler and its
+// collectors based on the config. This function also start the default
```
```suggestion
// collectors based on the config. This function also starts the default
```

olivielpeau reviewed (dismissed):
Just a couple of nits, other than that looks good to me!

Also, could you update the PR description with the important fact that this PR also changes the default interval of the `host` metadata collector?

olivielpeau commented on releasenotes/notes/change-metadata-interval-0dc52511f61ebfd1.yaml:
```diff
@@ -0,0 +1,10 @@
+---
+features:
```
nit: I don't think this really qualifies as a `feature`, I'd put this under `enhancement` :)

olivielpeau reviewed (dismissed):
Just a nit on the release note, feel free to merge once addressed!
